### PR TITLE
command line api to control http api functionality

### PR DESCRIPTION
**Pull request by colombod**
This PR aims to implement the command line api so that the http api service is enabled according to the [readme](dotnet-interactive/CommandLine/readme.md) document.

The expected bahviour is sumamrised as 

| Command                                                      | HTTP behavior
|--------------------------------------------------------------|--------------------------------
dotnet interactive http                                        | Enabled on auto-selected port
dotnet interactive http --http-port *                          | Enabled on auto-selected port
dotnet interactive http --http-port 1234                       | Enabled on port 1234
dotnet interactive http --http-port-range 8001-9000            | Enabled on a port between 8001 and 9000
dotnet interactive stdio                                       | Disabled
dotnet interactive stdio --http-port *                         | Enabled on auto-selected port
dotnet interactive stdio --http-port 1234                      | Enabled on port 1234
dotnet interactive stdio --http-port-range 8001-9000           | Enabled on a port between 8001 and 9000
dotnet interactive jupyter                                     | Disabled
dotnet interactive jupyter --http-port *                       | UNSUPPORTED
dotnet interactive jupyter --http-port 1234                    | UNSUPPORTED
dotnet interactive jupyter --http-port-range 8001-9000         | Enabled on a port between 8001 and 9000

When installing `dotnet-interactive` as a Jupyter kernel, you can also configure the HTTP API's availability when the kernel is started, as follows: 

| Command                                                      | HTTP behavior of kernel 
|--------------------------------------------------------------|--------------------------------
dotnet interactive jupyter install                             | Disabled
dotnet interactive jupyter install --http-port-range 8001-9000 | Enabled on autoselected port between 8001 and 9000

### CODE REVIEW

File: dotnet-interactive/CommandLine/CommandLineParser.cs [hunk 1]
```diff
@@ -364,23 +376,31 @@ Command KernelServer()
...
+
             SetUpFormatters(frontendEnvironment);
-            
 
             kernel.DefaultKernelName = defaultKernelName;
-            var enableHttp = new SubmitCode("#!enable-http", compositeKernel.Name);
-            enableHttp.PublishInternalEvents();
-            compositeKernel.DeferCommand(enableHttp);
+
+            if (frontendEnvironment is BrowserFrontendEnvironment _)
```
Comment (jonsequitur) [inline dotnet-interactive/CommandLine/CommandLineParser.cs:397]: Why repeat the `if`?

File: dotnet-interactive/CommandLine/CommandLineParser.cs [hunk 2]
```diff
@@ -364,23 +376,31 @@ Command KernelServer()
...
 
             var kernel = compositeKernel
                          .UseDefaultMagicCommands()
                          .UseLog()
-                         .UseAbout()
-                         .UseHttpApi(startupOptions, httpProbingSettings);
-            
+                         .UseAbout();
+
+            if (frontendEnvironment is BrowserFrontendEnvironment _)
```
Comment (jonsequitur) [inline dotnet-interactive/CommandLine/CommandLineParser.cs:388]: The discard shouldn't be needed.

File: dotnet-interactive/CommandLine/ReplFrontendEnvironment.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.DotNet.Interactive.App.CommandLine
+{
+    public class ReplFrontendEnvironment : FrontendEnvironment
```
Comment (jonsequitur) [inline dotnet-interactive/CommandLine/ReplFrontendEnvironment.cs:6]: What's this intended to represent? I don't see the `stdio` as being related to a true REPL.

File: dotnet-interactive/CommandLine/CommandLineParser.cs (hunk 1 above)
Comment (colombod) [reply dotnet-interactive/CommandLine/CommandLineParser.cs]: consolidated the logic in a single if

File: dotnet-interactive/CommandLine/ReplFrontendEnvironment.cs (hunk 3 above)
Comment (colombod) [reply dotnet-interactive/CommandLine/ReplFrontendEnvironment.cs:6]: to be more explicit this is a Livy scenario.
What the user here is expecting is
 * same rich formatting as browser frontend environment
 * no http api enabled

File: dotnet-interactive/CommandLine/CommandLineParser.cs (hunk 2 above)
Comment (colombod) [reply dotnet-interactive/CommandLine/CommandLineParser.cs]: removed the discard operator

File: dotnet-interactive/CommandLine/ReplFrontendEnvironment.cs (hunk 3 above)
Comment (colombod) [reply dotnet-interactive/CommandLine/ReplFrontendEnvironment.cs:6]: Maybe this is a **RemotedFrontendEnvironment**

Comment (jonsequitur) [reply dotnet-interactive/CommandLine/ReplFrontendEnvironment.cs]: As far as I can tell this is still a Browser front-end, though with different capabilities.

Comment (colombod) [reply dotnet-interactive/CommandLine/ReplFrontendEnvironment.cs]: agreed

File: dotnet-interactive/CommandLine/CommandLineParser.cs [hunk 4]
```diff
@@ -340,51 +293,52 @@ Command StdIO()
...
-                        {
-                            return startStdIO(
-                                startupOptions,
-                                CreateKernel(options.DefaultKernel,
-                                    new BrowserFrontendEnvironment(), startupOptions, null), console);
-                        }
+
+                        return startStdIO(
+                            startupOptions,
+                            CreateKernel(options.DefaultKernel,
```
Comment (jonsequitur) [inline dotnet-interactive/CommandLine/CommandLineParser.cs:309]: Nit: Line breaks please.

File: dotnet-interactive/CommandLine/StdIOCommand.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            if (kernel is KernelBase kernelBase)
+            {
+                kernelBase.RegisterForDisposal(disposable);
+                kernelBase.RegisterForDisposal(server);
+            }
+            await server.Input.LastAsync();
+            return 0;
+        }
+
+        internal static StandardIOKernelServer CreateServer(IKernel kernel, IConsole console)
```
Comment (jonsequitur) [inline dotnet-interactive/CommandLine/StdIOCommand.cs:29]: I think the method should enforce `KernelBase`.

File: dotnet-interactive/CommandLine/CommandLineParser.cs (hunk 4 above)
Comment (colombod) [reply dotnet-interactive/CommandLine/CommandLineParser.cs:309]: sure

**Pull request merged**